cleverhans/utils.py

```diff
@@ -220,6 +220,97 @@ def grid_visual(data):
     return figure
 
 
+def get_logits_over_interval(sess, model, x, adv_x, x_data,
+                             min_epsilon=-10, max_epsilon=10,
+                             num_points=21):
+    """Get logits when the input is perturbed in an interval in adv direction.
+
+    Args:
+        sess: Tf session
+        model: Model for which we wish to get logits
+        x: Data tensor
+        adv_x: Tensor for adversarial perturbation of x
+        x_data: Numpy array corresponding to the data
+            slice used to generate plot
+        min_epsilon: Minimum value of epsilon over the interval
+        max_epsilon: Maximum value of epsilon over the interval
+        num_points: Number of points used to interpolate
+
+    Returns:
+        Numpy array containing logits
+
+    Raises:
+        ValueError if min_epsilon is larger than max_epsilon
+    """
+    import tensorflow as tf
+    if min_epsilon > max_epsilon:
+        raise ValueError('Minimum epsilon is less than maximum epsilon')
+
+    epsilon = min_epsilon
+    eta = adv_x - x
+    eta = tf.nn.l2_normalize(eta, dim=0)
+    for i, epsilon in enumerate(np.linspace(min_epsilon,
+                                            max_epsilon,
+                                            num_points)):
+        with sess.as_default():
+            adv_x_epsilon = x + eta * epsilon
+            logits = model.get_logits(adv_x_epsilon)
+            log_prob_adv = sess.run(logits, feed_dict={x: x_data})
+            if i == 0:
+                log_prob_adv_array = log_prob_adv
+            else:
+                log_prob_adv_array = np.vstack((log_prob_adv_array,
+                                                log_prob_adv))
+
+    return log_prob_adv_array
+
+
+def linear_extrapolation_plot(log_prob_adv_array, y, file_name,
+                              min_epsilon=-10, max_epsilon=10,
+                              num_points=21):
+    """Generate linear extrapolation plot.
+
+    Args:
+        log_prob_adv_array: Numpy array containing log probabilities
+        y: Tf placeholder for the labels
+        file_name: Plot filename
+        min_epsilon: Minimum value of epsilon over the interval
+        max_epsilon: Maximum value of epsilon over the interval
+        num_points: Number of points used to interpolate
+    """
+    import matplotlib.pyplot as plt
+
+    # Ensure interactive mode is disabled and initialize our graph
+    plt.ioff()
```

Comment:
why does interactive mode need to be off? can we implement this function without having side effects on other code that uses pyplot?

Reply:
Done.